### ISSUE TITLE
Deconstruct for DateTime,  DateTimeOffset, DateOnly, and TimeOnly

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#24601]
# Proposal 

Allow retrieving the group of date and time parts in one call. This will help with performance and code writing/reading clarity.

# API Proposal

```C#

namespace System
{
    public readonly partial struct DateTime     
    {
        public void Deconstruct(out int year, out int month, out int day) { }
        public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second) {  }
        public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out int millisecond, out int microsecond) { }
    }

    public readonly partial struct DateTimeOffset
    {
        public void Deconstruct(out int year, out int month, out int day, out TimeSpan offset) { }
        public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out TimeSpan offset) {  }
        public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out int millisecond, out int microsecond, out TimeSpan offset) { }
    }

    public readonly partial struct DateOnly
    {
        public void Deconstruct(out int year, out int month, out int day) { }
    }
    
    public readonly partial struct TimeOnly
    {
        public void Deconstruct(out int hour, out int minute, out int second) { }
        public void Deconstruct(out int hour, out int minute, out int second, out int millisecond, out int microsecond) { }
    }
}

```

# Code Sample 

```C#
    var (year, month, day) = DateTime.Now;
```

# Old description

_From @panost on January 7, 2018 13:42_

For example method https://github.com/dotnet/coreclr/blob/0efe34efa69dea7f9b94ddc8251810e0a264671c/src/mscorlib/shared/System/DateTime.cs#L847

could be made public and renamed to Deconstruct, so we can write

    var (year,month,day) = DateTime.Now;

The same pattern can be applied to DateTimeOffset and TimeSpan

_Copied from original issue: dotnet/coreclr#15776_

**Comment by ericstj** [comment]
Moving this to 6.0 as it's not going to meet the deadline for 5.0.  @jkotas are you still interested in this issue?

**Comment by jkotas** [comment]
I just moved the issue to the right repo two years ago. I am not attached to this issue otherwise.

**Comment by KoalaBear84** [comment]
Be sure to also check for DateOnly/TimeOnly: #49036

**Comment by panost** [comment]
@KoalaBear84 This issue (DateOnly/TimeOnly) is closed, I can't comment there

@ericstj Sorry I didn't read your question a year ago. Yes it will be nice to be added. The code is in this [method ](https://github.com/dotnet/runtime/blob/c1b4be36072d30baaad3b6c75549d4a1104fe063/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L896). Also, it will be nice to be shared with DateOnly (through a common static method accepting dayNumber?)

**Comment by ericstj** [comment]
@tarekgh @mattjohnsonpint: thoughts on this issue?

**Comment by ericstj** [comment]
Related: https://github.com/dotnet/runtime/issues/19397 which added Deconstruct for `DictionaryEntry` and `KeyValuePair<TKey,TValue>`

**Comment by tarekgh** [comment]
The ask is reasonable here mainly for writing less code getting the various parts of the date and time and as pointed before it will be more optimized compared requesting every part separately. We need to look more carefully at the proposal to ensure right signatures.

**Comment by mattjohnsonpint** [comment]
I agree with Tarek.  We should think more carefully about the method signatures.

FWIW, This was also dropped as feedback [in comments](https://devblogs.microsoft.com/dotnet/date-time-and-time-zone-enhancements-in-net-6/#comment-9428) on my blog post about DateOnly.

**Comment by IanMarteens** [comment]
The main reason for having a Deconstruct for DateTime/DateOnly is that calling Year, Month and Day in three different calls is very inefficient. There's a lot of repeated work. I agree that a deconstructor for DateTime could be confusing, but that's not the case with DateOnly.

**Comment by panost** [comment]
@IanMarteens 
> a deconstructor for DateTime could be confusing

DateTime already has a constructor with the exact same arguments

**Comment by tarekgh** [comment]
> DateTime already has a constructor with the exact same arguments

It will be not obvious the deconstructor will always match the constructor parameters. I can see users can easily try to get the time parts only and can run into such problem. Imagine someone calling DateTime.Now.TimeOfDay and then trying to deconstruct that.
Deconstructor of DateOnly and TimeOnly can be reasonable to have though.

**Comment by julealgon** [comment]
> It will be not obvious the deconstructor will always match the constructor parameters.

@tarekgh does it need to be 100% evident, though? 

I was initially skeptical of having these deconstructors added since we are dealing with a lot of raw `int` variables that don't have inherent semantic value to distinguish themselves from one another.

However, after doing a quick test locally, I can see that VS is smart enough to provide proper deconstructor documentation when hovering over the assignment, which IMHO should be enough in case you don't remember exactly what the constructor parameter order is at that time (I don't think anyone should be required to memorize those, anyways):

![image](https://user-images.githubusercontent.com/461579/149028659-2202b872-adc6-4f79-b08c-d857831e8b25.png)

If such documentation was not provided to the user, then I'd agree with you this would be a bit rough.

I was surprised to see the documentation showing even for more obscure scenarios, like this with a `foreach` variable deconstruction:
![image](https://user-images.githubusercontent.com/461579/149029242-5ebf2f70-95bd-4494-b81e-a63b7bf0c28f.png)

**Comment by tarekgh** [comment]
@julealgon thanks for the feedback. 

How you think can get the `hour`, `minute`, and `second`? or anyone who wants to get these values must get the date parts too?

**Comment by panost** [comment]
@julealgon Yes, good point, also if in a rare occasion someone does ignore all the red flags, it is a first-run bug, it will be detected and corrected on the first run, I mean I can't imagine a code-base where this is a longstanding bug, someone after 6 months to discover that this piece of code actually prints years instead of hours!

Too much talk, about renaming and exposing a simple method that already exist, I will not pursue this issue any further, since it is too easy to write an extension method, that does exact;y this.
Too bad because the new coders will still use the sub-optimal Year, Month, Day properties. Also DateTime is the perfect example to explain what Deconstructing is, instead of examples of tuples copying fields.

@tarekgh Do you mean the time part of a DateTime (TimeOnly) or hours, minutes and seconds since AD ? Just joking! 
The most obvious path, is to use the Deconstructor of TimeOnly. You can add a sentence in DateTime docs, describing that path

```csharp
var (hours,minutes,_) = dateTime.TimeOnly;
```

**Comment by julealgon** [comment]
> @julealgon thanks for the feedback.
> 
> How you think can get the `hour`, `minute`, and `second`? or anyone who wants to get these values must get the date parts too?

I'm of the opinion that we shouldn't necessarily force people to get these values in any particular way, and that we could just add multiple deconstructors for multiple purposes.

For instance, since `DateTime` has both a `(year, month, day)` and a `(year, month, day, hour, minute, second)`, having both possibilities as deconstructors makes total sense to me. This allows one to do it like:
```csharp
var (_, _, _, hour, minute, second) = DateTime.Now;
```

The above might be considered a bit clunky (for the particular case of getting "only time information"), so alongside it, I think what would make the most sense to have a deconstructor added to `TimeSpan` so that one can deconstruct just the time components using:
```csharp
var (hour, minute, second) = DateTime.Now.TimeOfDay;
```

Or, using a deconstructor on `TimeOnly`, using:
```csharp
var (hour, minute, second) = TimeOnly.FromDateTime(DateTime.Now);
```

Each option makes sense in isolation so I think they should all be available.

> Too much talk, about renaming and exposing a simple method that already exist, I will not pursue this issue any further, since **it is too easy to write an extension method, that does exact;y this**.

@panost keep in mind that one of the reasons this request is being made is regarding performance. While it does help from a code readability standpoint to have the deconstructors and a single tuple to get the various values, accessing some of those properties on `DateTime` has a performance implication as well. As it stands, it is actually not possible to mitigate this performance cost via an external extension method, since you don't have access to the internals of the object to optimize this particular part. So even if you create the extension deconstructor, it will serve only the readability aspect, but not the performance one.

**Comment by panost** [comment]
The extensions I wrote are performant, they only need the DataTime.Ticks value. No other internal method of DataTime is needed.
For example:
```csharp
public static void Deconstruct( this DateTime dt, out int year, out int month, out int day.....
  int days = (int)( dt.Ticks / TimeSpan.TicksPerDay );
  GetDateParts( days, out year, out month, out _, out day );
  ...
```
the GetDateParts method, is a static equivalent of [DateTime.GetDate](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L852).

**Comment by tannergooding** [comment]
@tarekgh, is this something that's ready for review and if not, what's blocking it?

It sounded like you were in favor above, let me know if that isn't the case.

**Comment by tarekgh** [comment]
I think the proposal need to be updated a little to be ready for review. Mainly, the deconstruct or need to have all DateTime parts (including year, month, day...). The thing we need to think more about if we need to include milli and micro seconds too? 

Also, I am in favor adding deconstructor for DateOnly and TimeOnly too.

**Comment by Clockwork-Muse** [comment]
> The thing we need to think more about if we need to include milli and micro seconds too?

A tick is smaller than a microsecond.
Also, there's no constructor that takes anything smaller than a millisecond, so you couldn't roundtrip back through a constructor.

**Comment by tarekgh** [comment]
> Also, there's no constructor that takes anything smaller than a millisecond, so you couldn't roundtrip back through a constructor.

Not true. https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/DateTime.cs,568

**Comment by Clockwork-Muse** [comment]
Ah.  Which is new in 7, and the documentation shows 6 currently (because 7 is in preview).

**Comment by panost** [comment]
Who is the author here? Perhaps I am or is @jkotas ? Anyway I don't know what I am supposed to do. As I already said

> Too much talk, about renaming and exposing a simple method that already exist

with this addition
and add some deconstructors overloads that pair with the type's constructors
It is already a five years old issue

**Comment by tarekgh** [comment]
@panost you are the author of this issue. 

> Too much talk, about renaming and exposing a simple method that already exist

Discussion is always beneficiary. and the whole thing is pending on what I mentioned in the comment https://github.com/dotnet/runtime/issues/24601#issuecomment-1242548779. I am waiting to get some feedback if we include the smaller units in the deconstructors or not. Other than that, it is ready.

**Comment by IanMarteens** [comment]
> Discussion is always beneficiary. and the whole thing is pending on what I mentioned in the comment [#24601 (comment)](https://github.com/dotnet/runtime/issues/24601#issuecomment-1242548779). I am waiting to get some feedback if we include the smaller units in the deconstructors or not. Other than that, it is ready.

I wouldn't include smaller parts, unless the rule of matching each constructor with a constructor were written in stone. Personally, I don't see any use for extracting all those parts at once. Yes, there are performance improvements but, is there any practical use for that? I would stick to yMd and yMdhms. As a matter of fact, the whole issue with DateTime is less pressing now that we have DateOnly and TimeOnly, and proper support for them in EF7.

My two cents, of course.

**Comment by mattjohnsonpint** [comment]
Perhaps we only need a deconstructor on `DateOnly`.  Computing year, month, and day are the more expensive parts where having a deconstructor to do it in one pass would be most useful.  Calculating time components is much more straightforward arithmetic.

**Comment by tarekgh** [comment]
> Perhaps we only need a deconstructor on DateOnly

What about the scenario of DateTime users? they should convert DateTime to DateOnly and then deconstruct it? I think DateTime scenarios are more popular so far than DateOnly.

**Comment by mattjohnsonpint** [comment]
Fair point.  Then also for `DateTimeOffset`?

**Comment by panost** [comment]
> I don't see any use for extracting all those parts at once

[Here for example](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Date.R.cs,22) or writing a custom DateTime formatter

**Comment by IanMarteens** [comment]
> > I don't see any use for extracting all those parts at once
> 
> [Here for example](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Date.R.cs,22) or writing a custom DateTime formatter

Good point. Then, we should go all the way. Not necessarily one deconstructor for each constructor, but at least these three ones: year-month-day, year-month-day-hour-minute-second and would-you-give-all-you-have.

**Comment by tarekgh** [comment]
What you think about the following proposal? 

```C#

namespace System
{
    public readonly partial struct DateTime     
    {
        public void Deconstruct(out int year, out int month, out int day) { }
        public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second) {  }
        public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out int millisecond, out int microsecond) { }
    }

    public readonly partial struct DateOnly
    {
        public void Deconstruct(out int year, out int month, out int day) { }
    }
    
    public readonly partial struct TimeOnly
    {
        public void Deconstruct(out int hour, out int minute, out int second) { }
        public void Deconstruct(out int hour, out int minute, out int second, out int millisecond, out int microsecond) { }
    }
}

```

**Comment by IanMarteens** [comment]
> What you think about the following proposal?
> 

Perfect!

**Comment by panost** [comment]
for DateTimeOffset you will need to add the offset in minutes in every method

```csharp
public readonly partial struct DateTimeOffset
{
        public void Deconstruct(out int year, out int month, out int day,out short offsetMinutes) { }
        public void Deconstruct(out int year, out int month, out int day,out short offsetMinutes, out int hour, out int minute, out int second) {  }
        public void Deconstruct(out int year, out int month, out int day,out short offsetMinutes, out int hour, out int minute, out int second, out int millisecond, out int microsecond) { }
}
```

**Comment by tarekgh** [comment]
@panost good point. Why return offsetMinutes and not TimeSpan? constructors take TimeSpan and not offset minutes.

**Comment by Clockwork-Muse** [comment]
... especially since, historically, there have been sub-minute offsets (usually when a timezone transitioned off a local-solar-noon zone).

Although we don't have a way to get the seconds from `TimeZoneInfo`...

**Comment by tarekgh** [comment]
@Clockwork-Muse in DateTimeOffset we don't allow sub-minutes offsets. https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.cs,956.

My question is more about is it useful for users to get the value as TimeSpan or as minutes count. Also, having the constructors are taking TimeSpan is another factor I prefer Having TimeSpan.

**Comment by tarekgh** [comment]
Thinking more, I am seeing it is better to deconstruct using TimeSpan. This will make it consistent with the constructor. Also, we have another issue https://github.com/dotnet/runtime/issues/52081 which will expose `TotalOffsetMinutes` from DateTimeOffset if anyone wants to get this specific value. 

@panost let me know what you think about that.

**Comment by panost** [comment]
I vote for deconstructors that take the most optimal path. I always imagine a use case of serializing (deconstructing) millions of DateTimeOffset fields. So offsetInMinutes is just a copy of the private field. If you want a TimeSpan Offset you can always use the already available [Offset](https://source.dot.net/System.Private.CoreLib/R/1bc976bb59f3dd6e.html) property

But this is not a very strong opinion. If you leave the decosntructors as is (without time offset), it is also Ok with me, if #52081 is available. Adding TimeSpan you add [one (or three?) unnecessary multiplications](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/TimeSpan.cs,262a2ffd9cece820) for each call

**Comment by tarekgh** [comment]
Think about it more from the scenario point of view. When users really want to get the value as TimeSpan and when they need to get it as minutes count. In mainstream cases TimeSpan would be more useful. That is why I prefer to keep the TimeSpan.

**Comment by panost** [comment]
Yes I agree with that view, orthogonal design is a strong point, I mean keeping the arguments (and overloads) that the constructors have.

**Comment by tarekgh** [comment]
I marked the issue ready for the design review.

**Comment by terrajobst** [comment]
* We should consider having "simple" deconstructors that return the logical components
    - `DateTime` -> `(DateOnly, TimeOnly)`
    - `DateTimeOffset` -> `(DateOnly, TimeOnly, TimeSpan)`
* Conversely, we should consider adding them as constructors
    - `DateTime(DateOnly, TimeOnly)`
    - `DateTimeOffset(DateOnly, TimeOnly, TimeSpan)`
* We should give milliseconds without microseconds

```C#
namespace System;

public readonly partial struct DateTime
{
    public DateTime(DateOnly date, TimeOnly time);
    public DateTime(DateOnly date, TimeOnly time, DateTimeKind kind);
    public void Deconstruct(out DateOnly date, out TimeOnly time);
    public void Deconstruct(out int year, out int month, out int day);
    //public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second);
    //public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out int millisecond, out int microsecond);
}

public readonly partial struct DateTimeOffset
{
    public DateTimeOffset(DateOnly date, TimeOnly time, TimeSpan offset);
    public void Deconstruct(out DateOnly date, out TimeOnly time, out TimeSpan offset);
    // public void Deconstruct(out int year, out int month, out int day, out TimeSpan offset);
    // public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out TimeSpan offset);
    // public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second, out int millisecond, out int microsecond, out TimeSpan offset);
}

public readonly partial struct DateOnly
{
    public void Deconstruct(out int year, out int month, out int day);
}

public readonly partial struct TimeOnly
{
    public void Deconstruct(out int hour, out int minute);
    public void Deconstruct(out int hour, out int minute, out int second);
    public void Deconstruct(out int hour, out int minute, out int second, out int millisecond);
    public void Deconstruct(out int hour, out int minute, out int second, out int millisecond, out int microsecond);
}
```

**Comment by deeprobin** [comment]
@tarekgh 

Since I have already addressed some changes in the classes `DateTime`, `DateOnly` & `TimeOnly` so far, I would very much like to face this API.

If that goes okay, I would be happy if someone of you assigns me.

**Comment by tarekgh** [comment]
Thanks @deeprobin for willing to help with the implementation.

**Comment by svick** [comment]
> ```cs
>     public void Deconstruct(out DateOnly date, out TimeOnly time);
>     public void Deconstruct(out int year, out int month, out int day);
> ```

When `var (year, month, day) = dateTime;` works, is it a problem that `var (year, month) = dateTime;` does the wrong thing? Especially since cutting off the smaller parts works for deconstructing `TimeOnly`.

Though it should be fairly obvious what is wrong because of the types involved. And I don't have a better suggestion.

**Comment by panost** [comment]
@terrajobst  I can't imagine a use case for 
```csharp
 public void Deconstruct(out DateOnly date, out TimeOnly time);
````
I mean, where you will need both of them other than in [the two use cases](https://github.com/dotnet/runtime/issues/24601#issuecomment-1309925264) I already gave for 
```csharp
public void Deconstruct(out int year, out int month, out int day, out int hour, out int minute, out int second);
```
also the ctor of DateTime should also include Calendar
```csharp
public DateTime(DateOnly date, TimeOnly time, Calendar calendar, DateTimeKind kind)
```

**Comment by tarekgh** [comment]
> I mean, where you will need both of them (date and time) other than in https://github.com/dotnet/runtime/issues/24601#issuecomment-1309925264 I already gave for

You already gave the example for a scenario need the date and time parts. I can imagine other scenarios like scheduling calendars which will need to extract the date and time. We have avoided having a deconstruct which returns many outputs in this proposal. Also, talking logically, DateTime is really a Date and Time. deconstructing it to these parts makes a lot of sense. 

> also the ctors of DateTime from DateOnly, TimeOnly should include another overload with Calendar

Note, we have added the new constructors only because we added the deconstructors. If we find any scenario in the future that needs this constructor, we can consider adding it.
